// @tinyspeck/slack-desktop-utils/lib/index.js
const binary = require('@tinyspeck/node-pre-gyp');
const path = require('path');
const binding_path = binary.find(
  path.resolve(path.join(__dirname, '../package.json'))
);
const bindings = require(binding_path);

const platformOnly =
  (platform) =>
  (fn) =>
  (...args) => {
    if (process.platform !== platform) {
      throw new Error('Method not implemented');
    }
    return fn(...args);
  };

const registryListenerMap =
  process.platform === 'win32' ? new Map() : undefined;

const findRegistryListener = (key, callback) => {
  if (!registryListenerMap.has(key)) return [-1];
  const listeners = registryListenerMap.get(key);
  const idx = listeners.findIndex((listener) => listener.callback === callback);
  return idx > -1 ? [idx, listeners[idx]] : [idx];
};

bindings.systemPreferences.addRegistryListener = platformOnly('win32')(
  (hKey, subKey, callback) => {
    const key = `${hKey}\\${subKey}`;
    const [, listener] = findRegistryListener(key, callback);
    if (listener) return;

    const monitor = new bindings.systemPreferences.RegistryMonitor(
      hKey,
      subKey,
      callback
    );
    const listeners = registryListenerMap.get(key) || [];
    listeners.push({ callback, monitor });
    registryListenerMap.set(key, listeners);
  }
);

bindings.systemPreferences.removeRegistryListener = platformOnly('win32')(
  (hKey, subKey, callback) => {
    const key = `${hKey}\\${subKey}`;
    const [idx, listener] = findRegistryListener(key, callback);
    if (!listener) return;

    const listeners = registryListenerMap.get(key);
    listeners.splice(idx, 1);
    listener.monitor.close();

    if (listeners.length > 0) {
      registryListenerMap.set(key, listeners);
    } else {
      registryListenerMap.delete(key);
    }
  }
);

const WindowsNotifierState = {
  NotificationPlatformUnavailable: -2,
  Error: -1,
  Enabled: 0,
  DisabledForApplication: 1,
  DisabledForUser: 2,
  DisabledByGroupPolicy: 3,
  DisabledByManifest: 4,
};

const StartupTaskState = {
  Error: -1,
  Disabled: 0,
  DisabledByUser: 1,
  Enabled: 2,
  DisabledByPolicy: 3,
  EnabledByPolicy: 4,
};

const ActivationType = {
  Error: -1,
  StartupTask: 1020,
};

const AppRestartFlags = {
  NoCrash: 1 << 0,
  NoHang: 1 << 1,
  NoPatch: 1 << 2,
  NoReboot: 1 << 3,
};

function getWindowPositionAndDimensions() {
  const { screen } = require('electron');
  const x = this.getWindowXPosition();
  const y = this.getWindowYPosition();
  const width = this.getWindowWidth();
  const height = this.getWindowHeight();

  if (!x || !y || !width || !height) {
    return;
  }

  const rectangle = {
    x,
    y,
    width,
    height,
  };

  if (process.platform === 'win32') {
    // Convert virtual screen coordinates to scaled DIP coordinates.
    return screen.screenToDipRect(null, rectangle);
  }

  return rectangle;
}

bindings.huddles.WindowBorderPanel.prototype.getWindowPositionAndDimensions =
  getWindowPositionAndDimensions;

module.exports = {
  ...bindings,
  AppRestartFlags,
  WindowsNotifierState,
  StartupTaskState,
  ActivationType,
};
