// @tinyspeck/slack-desktop-utils/lib/index.js
const binary = require('@tinyspeck/node-pre-gyp');
const path = require('path');
const binding_path = binary.find(
  path.resolve(path.join(__dirname, '../package.json'))
);
const bindings = require(binding_path);

const platformOnly =
  (platform) =>
  (fn) =>
  (...args) => {
    if (process.platform !== platform) {
      throw new Error('Method not implemented');
    }
    return fn(...args);
  };

const registryListenerMap =
  process.platform === 'win32' ? new Map() : undefined;

const findRegistryListener = (key, callback) => {
  if (!registryListenerMap.has(key)) return [-1];
  const listeners = registryListenerMap.get(key);
  const idx = listeners.findIndex((listener) => listener.callback === callback);
  return idx > -1 ? [idx, listeners[idx]] : [idx];
};

bindings.systemPreferences.addRegistryListener = platformOnly('win32')(
  (hKey, subKey, callback) => {
    const key = `${hKey}\\${subKey}`;
    const [, listener] = findRegistryListener(key, callback);
    if (listener) return;

    const monitor = new bindings.systemPreferences.RegistryMonitor(
      hKey,
      subKey,
      callback
    );
    const listeners = registryListenerMap.get(key) || [];
    listeners.push({ callback, monitor });
    registryListenerMap.set(key, listeners);
  }
);

bindings.systemPreferences.removeRegistryListener = platformOnly('win32')(
  (hKey, subKey, callback) => {
    const key = `${hKey}\\${subKey}`;
    const [idx, listener] = findRegistryListener(key, callback);
    if (!listener) return;

    const listeners = registryListenerMap.get(key);
    listeners.splice(idx, 1);
    listener.monitor.close();

    if (listeners.length > 0) {
      registryListenerMap.set(key, listeners);
    } else {
      registryListenerMap.delete(key);
    }
  }
);

// Double-press accelerator - uses singleton in native layer for efficient single event tap/hook
// Store the native functions
const nativeRegister = bindings.accelerator.registerDoublePressShortcut;
const nativeUnregister = bindings.accelerator.unregisterDoublePressShortcut;
const nativeUnregisterAll = bindings.accelerator.unregisterAllDoublePressShortcuts;

/**
 * Register a callback for double-press of a modifier key, optionally followed by another key.
 *
 * Uses a singleton event tap (macOS) or keyboard hook (Windows) for efficiency -
 * multiple registrations share a single system hook.
 *
 * On macOS: Requires Accessibility permissions (use checkAccessibilityPermissions first)
 * On Windows: Uses low-level keyboard hooks (no special permissions required)
 *
 * @param {string} modifier - The modifier key: 'command'/'cmd', 'option'/'alt', 'control'/'ctrl', 'shift', 'win'
 * @param {Function} callback - Function called when the shortcut sequence is detected
 * @param {Object} options - Optional configuration
 * @param {number} options.threshold - Time threshold in ms for double-press (default: 300)
 * @param {string} options.finalKey - Optional final key to complete the sequence (e.g., 'e', 'z', 'shift')
 * @returns {string} Listener ID for unregistration
 *
 * @example
 * // Simple double-press: Cmd+Cmd
 * accelerator.registerDoublePressShortcut('command', callback);
 *
 * // With final key: Cmd+Cmd+E
 * accelerator.registerDoublePressShortcut('command', callback, { finalKey: 'e' });
 *
 * // With modifier as final key: Ctrl+Ctrl+Shift
 * accelerator.registerDoublePressShortcut('control', callback, { finalKey: 'shift' });
 */
bindings.accelerator.registerDoublePressShortcut = (modifier, callback, options = {}) => {
  const threshold = options.threshold ?? 300;
  const finalKey = options.finalKey ?? '';
  return nativeRegister(modifier, callback, threshold, finalKey);
};

/**
 * Unregister a previously registered double-press shortcut.
 *
 * @param {string} listenerId - The ID returned from registerDoublePressShortcut
 * @returns {boolean} true if successfully unregistered, false if not found
 */
bindings.accelerator.unregisterDoublePressShortcut = (listenerId) => {
  return nativeUnregister(listenerId);
};

/**
 * Unregister all double-press shortcuts.
 */
bindings.accelerator.unregisterAllDoublePressShortcuts = () => {
  nativeUnregisterAll();
};

const WindowsNotifierState = {
  NotificationPlatformUnavailable: -2,
  Error: -1,
  Enabled: 0,
  DisabledForApplication: 1,
  DisabledForUser: 2,
  DisabledByGroupPolicy: 3,
  DisabledByManifest: 4,
};

const StartupTaskState = {
  Error: -1,
  Disabled: 0,
  DisabledByUser: 1,
  Enabled: 2,
  DisabledByPolicy: 3,
  EnabledByPolicy: 4,
};

const ActivationType = {
  Error: -1,
  StartupTask: 1020,
};

const AppRestartFlags = {
  NoCrash: 1 << 0,
  NoHang: 1 << 1,
  NoPatch: 1 << 2,
  NoReboot: 1 << 3,
};

function getWindowPositionAndDimensions() {
  const { screen } = require('electron');
  const x = this.getWindowXPosition();
  const y = this.getWindowYPosition();
  const width = this.getWindowWidth();
  const height = this.getWindowHeight();

  if (!x || !y || !width || !height) {
    return;
  }

  const rectangle = {
    x,
    y,
    width,
    height,
  };

  if (process.platform === 'win32') {
    // Convert virtual screen coordinates to scaled DIP coordinates.
    return screen.screenToDipRect(null, rectangle);
  }

  return rectangle;
}

bindings.huddles.WindowBorderPanel.prototype.getWindowPositionAndDimensions =
  getWindowPositionAndDimensions;

const acceleratorComboKeyState = new Map();
const acceleratorCombos = new Map();
let acceleratorNextComboId = 1;
const acceleratorComboDownState = new Map();
let acceleratorNativeStarted = false;

function acceleratorIsComboDown(combo) {
  for (const key of combo.keys) {
    if (!acceleratorComboKeyState.get(key)) return false;
  }
  return true;
}

function invokeAcceleratorCallback(callback, callbackName) {
  try {
    callback();
  } catch (error) {
    console.error(`accelerator ${callbackName} callback threw`, error);
  }
}

function assertAcceleratorInteger(value, fieldName) {
  if (!Number.isInteger(value)) {
    throw new TypeError(`${fieldName} must be an integer`);
  }
}

function normalizeAcceleratorComboKeys(keys) {
  if (!Array.isArray(keys) || keys.length === 0) {
    throw new Error(
      'accelerator.registerCombo(combo, callbacks): combo.keys must be a non-empty array of key codes'
    );
  }

  const seen = new Set();
  return keys.map((key, index) => {
    assertAcceleratorInteger(key, `combo.keys[${index}]`);
    if (seen.has(key)) {
      throw new Error(
        'accelerator.registerCombo(combo, callbacks): combo.keys must not contain duplicates'
      );
    }
    seen.add(key);
    return key;
  });
}

function validateAcceleratorComboCallbacks(callbacks) {
  if (callbacks === undefined) return {};
  if (!callbacks || typeof callbacks !== 'object') {
    throw new TypeError(
      'accelerator.registerCombo(combo, callbacks): callbacks must be an object'
    );
  }

  const {
    onKeyDown,
    onKeyUp,
  } = callbacks;

  if (onKeyDown !== undefined && typeof onKeyDown !== 'function') {
    throw new TypeError('callbacks.onKeyDown must be a function when provided');
  }
  if (onKeyUp !== undefined && typeof onKeyUp !== 'function') {
    throw new TypeError('callbacks.onKeyUp must be a function when provided');
  }

  return callbacks;
}

function validateAcceleratorRegisterComboArgs(combo, callbacks) {
  if (!combo || typeof combo !== 'object' || Array.isArray(combo)) {
    throw new TypeError(
      'accelerator.registerCombo(combo, callbacks): combo must be an object'
    );
  }

  return {
    keys: normalizeAcceleratorComboKeys(combo.keys),
    callbacks: validateAcceleratorComboCallbacks(callbacks),
  };
}

function acceleratorInternalOnKeyDown(keyCode) {
  acceleratorComboKeyState.set(keyCode, true);

  for (const combo of Array.from(acceleratorCombos.values())) {
    if (!acceleratorIsComboDown(combo)) continue;

    const wasDown = acceleratorComboDownState.get(combo.id) || false;
    if (wasDown) continue;

    acceleratorComboDownState.set(combo.id, true);
    if (typeof combo.callbacks.onKeyDown === 'function') {
      invokeAcceleratorCallback(combo.callbacks.onKeyDown, 'onKeyDown');
    }
  }
}

function acceleratorInternalOnKeyUp(keyCode) {
  acceleratorComboKeyState.set(keyCode, false);

  for (const combo of Array.from(acceleratorCombos.values())) {
    const wasDown = acceleratorComboDownState.get(combo.id) || false;
    if (!wasDown) continue;

    if (!acceleratorIsComboDown(combo)) {
      acceleratorComboDownState.set(combo.id, false);
      if (typeof combo.callbacks.onKeyUp === 'function') {
        invokeAcceleratorCallback(combo.callbacks.onKeyUp, 'onKeyUp');
      }
    }
  }
}

function acceleratorEnsureNativeStarted(keys) {
  if (acceleratorNativeStarted || !acceleratorNativeStart) return;
  acceleratorNativeStart(
    { keys },
    acceleratorInternalOnKeyDown,
    acceleratorInternalOnKeyUp
  );
  acceleratorNativeStarted = true;
}

function acceleratorStopNativeIfIdle() {
  if (acceleratorCombos.size > 0 || !acceleratorNativeStarted) return;
  if (acceleratorNativeStop) acceleratorNativeStop();
  acceleratorNativeStarted = false;
  acceleratorComboKeyState.clear();
}

function acceleratorUnregisterComboById(id) {
  assertAcceleratorInteger(id, 'combo id');

  const entry = acceleratorCombos.get(id);
  if (!entry) return false;

  acceleratorCombos.delete(id);
  acceleratorComboDownState.delete(id);
  acceleratorStopNativeIfIdle();
  return true;
}

function acceleratorUnregisterAllCombosInternal() {
  acceleratorCombos.clear();
  acceleratorComboDownState.clear();
  acceleratorStopNativeIfIdle();
}

const acceleratorNativeStart = bindings.accelerator._startComboHook;
const acceleratorNativeStop = bindings.accelerator._stopComboHook;

bindings.accelerator.registerCombo = function (combo, callbacks) {
  const normalized = validateAcceleratorRegisterComboArgs(combo, callbacks);
  const hasDifferentCombo = Array.from(acceleratorCombos.values()).some(
    (comboEntry) =>
      comboEntry.keys.length !== normalized.keys.length ||
      comboEntry.keys.some((key, index) => key !== normalized.keys[index])
  );
  if (hasDifferentCombo) {
    throw new Error(
      'accelerator.registerCombo(): only one unique combo is supported at a time'
    );
  }
  const id = acceleratorNextComboId++;
  const entry = {
    id,
    keys: normalized.keys,
    callbacks: normalized.callbacks,
  };
  acceleratorEnsureNativeStarted(normalized.keys);
  try {
    acceleratorCombos.set(id, entry);
  } catch (err) {
    acceleratorCombos.delete(id);
    acceleratorComboDownState.delete(id);
    acceleratorStopNativeIfIdle();
    throw err;
  }

  return id;
};

bindings.accelerator.unregisterCombo = function (id) {
  return acceleratorUnregisterComboById(id);
};

bindings.accelerator.unregisterAllCombos = function () {
  acceleratorUnregisterAllCombosInternal();
};
delete bindings.accelerator._startComboHook;
delete bindings.accelerator._stopComboHook;

if (!bindings.clipboard.paste) {
  bindings.clipboard.paste = async function () {
    throw new Error('Method not implemented');
  };
}

module.exports = {
  ...bindings,
  AppRestartFlags,
  WindowsNotifierState,
  StartupTaskState,
  ActivationType,
};
